core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -775,10 +775,14 @@ object IO {
   final def require[E, A](error: E): IO[E, Option[A]] => IO[E, A] =
     (io: IO[E, Option[A]]) => io.flatMap(_.fold[IO[E, A]](IO.fail[E, A](error))(IO.now[E, A]))
 
-  def forkAll[E2](l: List[IO[E2, Unit]]): IO[E2, Unit] = l match {
-    case Nil     => IO.unit[E2]
-    case x :: xs => x.fork *> forkAll(xs)
-  }
+  final def forkAll[E, E2, A](as: List[IO[E, A]]): IO[E2, Fiber[E, List[A]]] =
+    as.foldRight(IO.point[E2, Fiber[E, List[A]]](Fiber.point(List.empty))) {
+      case (a, fiberAs) =>
+        for {
+          fiberA   <- a.fork
+          fiberAs0 <- fiberAs.map(_.zipWith(fiberA)((as, a) => a :: as))
```

Comment:
Can you rewrite this using flatMap so it becomes tail recursive?

Reply:
I need to rewrite the for-comprehension so `zipWith` is the last call in the for comprehension or the entire `foldRight`?

Reply:
In this case, I just mean without the final useless ‘map’ that Scala’s desugaring will insert.

Reply:
We should consider adding oleg-py/better-monadic-for for stuff like that.